core/src/main/java/org/apache/iceberg/puffin/PuffinWriter.java

```diff
@@ -0,0 +1,189 @@
+    this.defaultBlobCompression = defaultBlobCompression;
+  }
+
+  @Override
+  public void add(Blob blob) {
+    Preconditions.checkNotNull(blob, "blob is null");
+    try {
+      writeHeaderIfNeeded();
+      long fileOffset = outputStream.getPos();
+      PuffinCompressionCodec codec = MoreObjects.firstNonNull(blob.requestedCompression(), defaultBlobCompression);
+      ByteBuffer rawData = PuffinFormat.compress(codec, blob.blobData());
+      int length = rawData.remaining();
+      writeFully(rawData);
+      writtenBlobsMetadata.add(new BlobMetadata(blob.type(), blob.inputFields(), fileOffset, length,
+          codec.codecName(), blob.properties()));
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public Metrics metrics() {
+    return new Metrics();
+  }
+
+  @Override
+  public long length() {
+    return fileSize();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (!finished) {
+      finish();
+    }
+
+    outputStream.close();
+  }
+
+  private void writeHeaderIfNeeded() throws IOException {
+    if (headerWritten) {
+      return;
+    }
+
+    this.outputStream.write(MAGIC);
+    this.headerWritten = true;
+  }
+
+  public void finish() throws IOException {
+    checkNotFinished();
+    writeHeaderIfNeeded();
+    if (footerSize.isPresent()) {
+      throw new IllegalStateException("footerSize already set");
+    }
+
+    long footerOffset = outputStream.getPos();
+    writeFooter();
+    this.footerSize = Optional.of(Math.toIntExact(outputStream.getPos() - footerOffset));
+    this.fileSize = Optional.of(outputStream.getPos());
+    this.finished = true;
+  }
+
+  private void writeFooter() throws IOException {
+    FileMetadata fileMetadata = new FileMetadata(writtenBlobsMetadata, properties);
+    ByteBuffer footerJson = ByteBuffer.wrap(FileMetadataParser.toJson(fileMetadata, false)
+        .getBytes(StandardCharsets.UTF_8));
+    ByteBuffer footerPayload = PuffinFormat.compress(footerCompression, footerJson);
+    outputStream.write(MAGIC);
+    int footerPayloadLength = footerPayload.remaining();
+    writeFully(footerPayload);
+    PuffinFormat.writeIntegerLittleEndian(outputStream, footerPayloadLength);
+    writeFlags();
+    outputStream.write(MAGIC);
+  }
+
+  private void writeFlags() throws IOException {
+    Map<Integer, List<Flag>> flagsByByteNumber = fileFlags().stream()
+        .collect(Collectors.groupingBy(Flag::byteNumber));
+    for (int byteNumber = 0; byteNumber < PuffinFormat.FOOTER_STRUCT_FLAGS_LENGTH; byteNumber++) {
+      int byteFlag = 0;
+      for (Flag flag : flagsByByteNumber.getOrDefault(byteNumber, ImmutableList.of())) {
+        byteFlag |= 0x1 << flag.bitNumber();
+      }
+      outputStream.write(byteFlag);
+    }
+  }
+
+  private void writeFully(ByteBuffer buffer) throws IOException {
+    WritableByteChannel channel = Channels.newChannel(outputStream);
```

Comment:
I think this needs to be wrapped in try-with-resources because `Channel` is `Closeable`.

Reply:
per `Channels.newChannel` docs:

>  Closing the channel will in turn cause the stream to be closed.

this is not what we want here. 

is there a better way to write a `ByteBuffer` to an `OutputStream`?

Reply:
Got it. How about we wrap the stream and keep a reference to the channel instead?

Reply:
if it is only about an IDE false positive warning about a Closeable not being closed, i think the right recourse is to add a suppression.

in fact, i would prefer to avoid Channel at all, because Channels interact with interrupts in a manner that's not intuitive to me (eg https://github.com/delta-io/delta/commit/7a3f1e8ec626e80880d524c2b897a969c8b4d63a).

Please let me know which approach would look awesome from Iceberg maintainers' point of view.

Reply:
If you'd like to avoid using Channel, that's fine with me. I just want to make sure we're not leaking resources and this uses a factory method so we aren't guaranteed to have a channel that doesn't require being closed.

Why don't you write a method to implement `writeFully` from a byte buffer in `IOUtil` and we can just call that instead?

Reply:
> I just want to make sure we're not leaking resources and this uses a factory method so we aren't guaranteed to have a channel that doesn't require being closed.

right

or, an implementation which is buffering, and we end up not writing (all) the data

> Why don't you write a method to implement writeFully from a byte buffer in IOUtil and we can just call that instead?

will do!

Reply:
fixed